tests and cleanup for python3 and multiprocessing

jhamman opened the pull request:
This PR fixes an issue with the RVIC multiprocessing formulation where errors were not raised by Pool workers.  This has been fixed and a regression test has been added.

fixes #50

jhamman commented:
@bartnijssen - thank you for the code review on this.  Here's a quick summary of where we're at:

- This is the current working branch: `jhamman:feature/fix_multiproc`
- This has been broken for a while and I haven't been convinced that it is broken because of the changes involved in making RVIC work with Python 3

#### `multiprocessing` workflow:
  - We're using `multiprocessing.Pool` and `multiprocessing.Pool.apply_async` to manage processes in [`parameters.py`](https://github.com/jhamman/RVIC/blob/feature/fix_multiproc/rvic/parameters.py#L36).
  - [Multiprocessing Pool documentation](https://docs.python.org/3.4/library/multiprocessing.html#multiprocessing.pool.Pool)
  - Using apply_async to allow passing multiple arguments to function (`gen_uh_run`)
  - We're using the `callback` method to return results of `gen_uh_run'
  - This callback method calls the [`store_result` function](https://github.com/jhamman/RVIC/blob/feature/fix_multiproc/rvic/parameters.py#L533) which appends the results of `gen_uh_run` to a `global` `list` `results`.
  - Actually, we're using a subclass of the `multiprocessing.Pool` called [`LoggingPool`](https://github.com/jhamman/RVIC/blob/feature/fix_multiproc/rvic/core/multi_proc.py#L40), which is defined in `multi_proc.py`.  This subclass is "supposed" to help return the full traceback from processes that error while in a subprocess.

The two tests that are failing are defined in [`test_mulit_proc.py`](https://github.com/jhamman/RVIC/blob/feature/fix_multiproc/tests/unit/test_multi_proc.py):
  - can be run with `py.test test_multi_proc.py`
 - `test_pool_callback` tests to make sure the global list is populated on callback.
  - `test_pool_raises` tests that exceptions raised in subprocesses are returned to the master processor.

That should be about it.  Please ask questions if anything doesn't make sense.  I'm wondering if I'm missing something in how the subclass of `Pool` is working or with the `callback` method.

bartnijssen commented:
Is multi-processing used in UW-Hydro:develop and if so is that one working?

jhamman commented:
Yes, it is used in UW-Hydro:develop.
No, it is not working in UW-Hydro:develop.
It is working in UW-Hydro:master.

jhamman commented:
@bartnijssen - this is ready for review

bartnijssen commented:
Most of the changes here do not appear to have anything to do with `multiprocessing`, but with changes from lists to dicts and from regular dicts to OrderDicts. Please do one of the following:

* change title of PR and description to describe the additional changes (this is probably the easiest)

or

* split this into more than one PR

The code itself looks fine. I am not looking for a lengthy description, but just a few lines and if possible a better title (not sure you can change the title of a PR, if not, leave it be).

jhamman commented on rvic/parameters.py:
```diff
@@ -349,6 +363,18 @@ def gen_uh_init(config_file):
+            print(dom_data['cell_ids'].shape)
+            print([domys[i], domxs[i]])
+
+            print(dict(cell_id=dom_data['cell_ids'][domys[i], domxs[i]]))
```
these print statements should be removed

jhamman commented:
@bartnijssen - I updated the title, but in fact, almost all the changes in this PR were required to get multiprocessing working again (including changes between lists and dicts).

So, a brief summary of the changes:
- changed `results` from `list` to `dict` to support easier collection of results from worker processes
- removed `logging` wrapper around `multiprocessing.Pool`.  This was causing more problems than it was worth.  We now raise a real python Exception if an error is raised in a worker process.
- fixed bug in `search_for_channel` that was found to be causing problems in multiprocessing level.
- skip cdo test on travis (this is barely a real test and could be removed all together).
